### PR TITLE
[DDP] Enable delayed all reduce in DDP to deprecate DelayAllReduceDDP

### PR DESCRIPTION
**Pull request by GD06**
Summary: Enable the functionality of delaying all reduce in DDP to deprecate DelayAllReduceDDP.

Test Plan:
# Unit Test

```
buck2 test mode/opt //hpc/torchrec/tests:data_parallel_test
```

Differential Revision: D43586898

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -528,6 +528,11 @@ class DistributedDataParallel(Module, Joinable):
                          >>> ...
                          >>> ddp_logging_data = model_DDP._get_ddp_logging_data()
                          >>> static_graph = ddp_logging_data.get("can_set_static_graph")
+        delay_all_reduce_named_params (list of tuple of str and torch.nn.Parameter): a list
+                    of named parameters whose all reduce will be delayed into the backward
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:532]: nit: will be delayed when the gradient of the parameter specified in ``param_to_hook_all_reduce`` is ready

Comment (zhaojuanmao) [reply torch/nn/parallel/distributed.py:532]: nit: could we explain a little bit about how the two arguments will work with other arguments?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -661,6 +684,20 @@ def __init__(
             os.environ.get("PYTORCH_DDP_USE_SIDE_STREAM", "1") == "1"
         )
 
+        # Initialize gradient buffers and register all reduce hook
+        self._grad_buffer = None
+        self._grad_views = []
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:689]: nit: self._delay_grad_buffer, self._delay_grad_views

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -661,6 +684,20 @@ def __init__(
...
+        self._use_vanilla_module = False
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:690]: nit: self._delay_all_reduce = False

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -661,6 +684,20 @@ def __init__(
...
+        if len(self._delay_all_reduce_params) != 0:
+            self._register_delay_all_reduce_hook(
+                bucket_cap_mb=bucket_cap_mb,
+                process_group=self.process_group,
+                param_to_hook_all_reduce=param_to_hook_all_reduce,
+                device_ids=device_ids,
+            )
+            if self._use_vanilla_module:
+                return
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:699]: nit: seems this is not needed?

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -691,6 +728,52 @@ def __init__(
...
+        # 2. Broadcast the parameters
+        detached_params = [p.detach() for p in self._delay_all_reduce_params]
+        dist._broadcast_coalesced(process_group, detached_params, bucket_cap_mb, 0)
+
+        # 3. Hook all reduce to the specified parameter
+        world_size = dist.get_world_size(process_group)
+
+        def _delayed_all_reduce(grad):
+            self._grad_buffer.div_(world_size)
+            _ = dist.all_reduce(self._grad_buffer, group=process_group, async_op=True)
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:754]: where to wait for the all_reduce to complete? are you assuming optimizers will just wait for their grads?

File: torch/nn/parallel/distributed.py [hunk 6]
```diff
@@ -691,6 +728,52 @@ def __init__(
...
+            self._grad_views.append(grad_view)
+            offset = offset + param.numel()
+
+        # 5. Check whether the all reduce of all params requiring grad is delayed.
+        for module_name, module in self.module.named_modules():
+            for param_name, param in module.named_parameters(recurse=False):
+                if param.requires_grad:
+                    full_name = f"{module_name}.{param_name}"
+                    if full_name not in self.parameters_to_ignore:
+                        return
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:774]: seems that it is not checking anything? it just simply return?

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -1190,6 +1284,9 @@ def forward(self, *inputs, **kwargs):
             output = _tree_unflatten_with_rref(
                 output_placeholders, treespec, output_is_rref
             )
+
+        # At the end of the forward pass, reset the grad buffer and grad views
+        self._clear_grad_buffer()
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:1289]: this can be removed?

File: torch/nn/parallel/distributed.py [hunk 8]
```diff
@@ -1085,10 +1168,21 @@ def _run_ddp_forward(self, *inputs, **kwargs):
...
+            self._grad_buffer.zero_()
+
     def forward(self, *inputs, **kwargs):
         with torch.autograd.profiler.record_function(
             "DistributedDataParallel.forward"
         ):
+            if self._use_vanilla_module:
+                output = self.module.forward(*inputs, **kwargs)
+                self._clear_grad_buffer()
+                return output
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:1184]: how about the parameters that are not managed by _delay_all_reduce_params? they should be managed by DDP reducer, for these parameters, we still need to execute the below codes for preparing DDP reducer?

File: torch/nn/parallel/distributed.py [hunk 9]
```diff
@@ -1085,10 +1168,21 @@ def _run_ddp_forward(self, *inputs, **kwargs):
             with self._inside_ddp_forward():
                 return self.module(*inputs, **kwargs)
 
+    def _clear_grad_buffer(self):
+        if self._grad_buffer is not None:
+            for index, param in enumerate(self._delay_all_reduce_params):
+                param.grad = self._grad_views[index]
+            self._grad_buffer.zero_()
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:1175]: just want to have a head up or you can add a comment here: "making param.grad points to the grad buffers before backward" is based on the assumption that the grad accumulation is done in place in autograd engine, for some edge cases, if the grad accumulation in autograd engine is not in place, then the param.grad and grad buffers are detached

**Review by zhaojuanmao — COMMENTED** [review]
Thanks for adding this! overall looks good, left some comments

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:532]: Thanks Yanli for suggestions! I revised comments to improve these.

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:689]: Done.

File: torch/nn/parallel/distributed.py (hunk 3 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:690]: The flag "use_vanilla_module" is used to flag whether all parameters are specified in delay_all_reduce_named_params. In this case, we don't need DDP and use the forward function of the self.module.

File: torch/nn/parallel/distributed.py (hunk 4 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:699]: When we are registering delay all reduce hook, we will check whether all parameters are specified in delay_all_reduce_named_params. If yes, we don't need DDP to construct reducer or other facilities.

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:754]: In our use case, we assume this happens before the final all reduce (DenseArch all reduce). When they are using the same PG, we don't have to explicitly wait for this all reduce. 

But I agree that this brings the risk of missing synchronization. Do you think I should comment here about our assumptions?

File: torch/nn/parallel/distributed.py (hunk 6 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:774]: If we have param that is not specified in parameters_to_ignore, we should not set _use_vanilla_module to True.

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:1289]: We need to manually set grad buffer of these delayed all reduce to zero, and rebuild grad views if optimizer.zero_grad sets them to None.

File: torch/nn/parallel/distributed.py (hunk 8 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:1184]: Yes. In the case where there are still params that are not managed by _delay_all_reduce_params, the flag _use_vanilla_module is False.

File: torch/nn/parallel/distributed.py (hunk 9 above)
Comment (GD06) [reply torch/nn/parallel/distributed.py:1175]: I added a comment here to clarify this implicit assumption.

**Review by zhaojuanmao — COMMENTED** [review]
thanks for explaining, still some minor comments

File: torch/nn/parallel/distributed.py [hunk 10]
```diff
@@ -1085,10 +1170,25 @@ def _run_ddp_forward(self, *inputs, **kwargs):
...
 
+    def _clear_grad_buffer(self):
+        # Making param.grad points to the grad buffers before backward is based on the
+        # assumption that the grad accumulation is done in place in autograd engine,
+        # for some edge cases, if the grad accumulation in autograd engine is not in
+        # place, then the param.grad and grad buffers are detached.
+        if self._delay_grad_buffer is not None:
+            for index, param in enumerate(self._delay_all_reduce_params):
+                param.grad = self._delay_grad_views[index]
+            self._delay_grad_buffer.zero_()
```
[the remark this answers is not in the document]
Comment (GD06) [reply torch/nn/parallel/distributed.py]: I changed the variable "_use_vanilla_module" to "_dar_all_params" (dar: delay-all-reduce) and added some comments to clarify code logics.

Also, thank you for your suggestions about grad_buffer! As zero_grad is a part of optimizer, we only need to rebuild param.grad when they are set to None. Thus I revised this function as follows:

We check whether the grad of all params is set to None. If yes, we will clear the whole grad buffer. If no, we will clear the grad of params individually. When iterating through self._delay_all_reduce_params, we only reset param.grad whose value is None.

**Comment by zhaojuanmao** [comment]
overall lgtm, thank you! seems lint runner has some errors, will approve once the CI is green

**Comment by GD06** [comment]
@zhaojuanmao  Hi Yanli, CI on GitHub turns to green. Could you please take another look?

**Review by zhaojuanmao — APPROVED** [review]
looks good, thanks!!

File: torch/nn/parallel/distributed.py [hunk 11]
```diff
@@ -661,6 +686,20 @@ def __init__(
             os.environ.get("PYTORCH_DDP_USE_SIDE_STREAM", "1") == "1"
         )
 
+        # Initialize gradient buffers and register all reduce hook
+        self._delay_grad_buffer = None
+        self._delay_grad_views = []
+        self._dar_all_params = False
```
[the remark this answers is not in the document]
Comment (GD06) [reply torch/nn/parallel/distributed.py:692]: Done.

**Comment by GD06** [comment]
Abandon this pull request as the corresponding Diff has been abandoned.
